app/src/main/java/org/schabi/newpipe/player/PlayerService.kt

```diff
@@ -0,0 +1,352 @@
+        mediaSession?.setActive(false)
+
+        // Should already be handled by NotificationUtil.cancelNotificationAndStopForeground() in
+        // NotificationPlayerUi, but let's make sure that the foreground service is stopped.
+        ServiceCompat.stopForeground(this, ServiceCompat.STOP_FOREGROUND_REMOVE)
+    }
+
+    /**
+     * Destroys the player and allows the player instance to be garbage collected. Sets the media
+     * session to inactive. Stops the foreground service and removes the player notification
+     * associated with it. Tries to stop the [PlayerService] completely, but this step will
+     * have no effect in case some service connection still uses the service (e.g. the Android Auto
+     * system accesses the media browser even when no player is running).
+     */
+    fun destroyPlayerAndStopService() {
+        if (DEBUG) {
+            Log.d(TAG, "destroyPlayerAndStopService() called")
+        }
+
+        cleanup()
+
+        // This only really stops the service if there are no other service connections (see docs):
+        // for example the (Android Auto) media browser binder will block stopService().
+        // This is why we also stopForeground() above, to make sure the notification is removed.
+        // If we were to call stopSelf(), then the service would be surely stopped (regardless of
+        // other service connections), but this would be a waste of resources since the service
+        // would be immediately restarted by those same connections to perform the queries.
+        stopService(Intent(this, PlayerService::class.java))
+    }
+
+    override fun attachBaseContext(base: Context?) {
+        super.attachBaseContext(AudioServiceLeakFix.preventLeakOf(base))
+    }
+
+    //endregion
+    //region Bind
+    override fun onBind(intent: Intent): IBinder? {
+        if (DEBUG) {
+            Log.d(
+                TAG,
+                (
+                    "onBind() called with: intent = [" + intent +
+                        "], extras = [" + intent.extras.toDebugString() + "]"
+                    )
+            )
+        }
+
+        if (BIND_PLAYER_HOLDER_ACTION == intent.action) {
+            // Note that this binder might be reused multiple times while the service is alive, even
+            // after unbind() has been called: https://stackoverflow.com/a/8794930 .
+            return mBinder
+        } else if (SERVICE_INTERFACE == intent.action) {
+            // MediaBrowserService also uses its own binder, so for actions related to the media
+            // browser service, pass the onBind to the superclass.
+            return super.onBind(intent)
+        } else {
+            // This is an unknown request, avoid returning any binder to not leak objects.
+            return null
+        }
+    }
+
+    class LocalBinder internal constructor(playerService: PlayerService) : Binder() {
+        private val playerService: WeakReference<PlayerService?>
```

Comment:
You can combine declaration and assignment.
```suggestion
        private val playerService: WeakReference<PlayerService> = WeakReference<PlayerService>(playerService)
```